cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/server.go

```diff
@@ -172,6 +201,130 @@ func (s *Server) GetAvailablePackages(ctx context.Context, request *corev1.GetAv
+	log.Infof("+GetPackageMeta()")
+
+	url, err := s.pullChartTarball(ctx, request.Package)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("Found chart url: [%s]", *url)
+
+	// unzip and untar .tgz file
+	meta, err := fetchMetaFromChartTarball(request.Package.Name, *url)
+	if err != nil {
+		return nil, err
+	}
+
+	return &corev1.GetPackageMetaResponse{
+		Meta: &corev1.GetPackageMetaResponse_PackageMeta{
+			Readme: meta[readme],
+		},
+	}, nil
+}
+
+func (s *Server) pullChartTarball(ctx context.Context, packageRef *corev1.PackageReference) (*string, error) {
+	client, err := s.GetClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceIfc := client.Resource(chartsResource).Namespace("default")
+
+	// see if we the chart already exists
+	chartList, err := resourceIfc.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO it'd be better if we could filter on server-side
+	for _, unstructuredChart := range chartList.Items {
+		chartName, found, err := unstructured.NestedString(unstructuredChart.Object, "spec", "chart")
+		if err == nil && found && chartName == packageRef.Name {
+			done, err := isChartPullComplete(&unstructuredChart)
+			if err != nil {
+				return nil, err
+			} else if done {
+				url, found, err := unstructured.NestedString(unstructuredChart.Object, "status", "url")
+				if err != nil || !found {
+					return nil, status.Errorf(codes.Internal, "expected field status.url not found on HelmChart: %v:\n%v", err, unstructuredChart)
+				}
+				log.Infof("Found existing HelmChart for: [%s]", packageRef.Name)
+				return &url, nil
+			}
+			// TODO waitUntilChartPullComplete
+		}
+	}
+
+	// did not find the chart, need to create
+	// see https://fluxcd.io/docs/components/source/helmcharts/
+	unstructuredChart := unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": fmt.Sprintf("%s/%s", fluxGroup, fluxVersion),
+			"kind":       fluxHelmChart,
+			"metadata": map[string]interface{}{
+				"generateName": fmt.Sprintf("%s-", packageRef.Name),
+			},
+			"spec": map[string]interface{}{
+				"chart": packageRef.Name,
+				"sourceRef": map[string]interface{}{
+					"name": packageRef.Repository.Name,
+					"kind": "HelmRepository",
+				},
+				"interval": "10m",
+			},
+		},
+	}
+
+	newChart, err := resourceIfc.Create(ctx, &unstructuredChart, metav1.CreateOptions{})
+	if err != nil {
+		log.Errorf("error creating chart: %v\n%v", err, unstructuredChart)
+		return nil, err
+	}
+
+	log.Infof("created chart: [%v]", newChart)
+
+	// wait until flux reconciles
+	watcher, err := resourceIfc.Watch(ctx, metav1.ListOptions{
+		ResourceVersion: newChart.GetResourceVersion(),
+	})
+	if err != nil {
+		log.Errorf("error creating watch: %v\n%v", err, unstructuredChart)
+		return nil, err
+	}
+
+	return waitUntilChartPullComplete(watcher)
+}
+
+func waitUntilChartPullComplete(watcher watch.Interface) (*string, error) {
+	ch := watcher.ResultChan()
```

Comment:
Nice - great use of both channel and the watcher to handle the async request. Note that this effectively makes `pullChartTarball` a blocking call, which is fine in it's current use (where it's only called in a context of one chart). Not for now, but in the future you might instead want to consider something like passing a results channel (of string urls) to pullChartTarball, so it returns immediately and you wait on the results channel at the call-site, which would mean you could call it for 20 different charts and just wait for the results to come in whatever order they happen to take, rather than serially.